File: fortress-crypto-android/src/main/kotlin/com/matrix/crypto/MatrixCryptoBridge.kt
package com.matrix.crypto

/**
 * Kotlin wrapper for the Matrix crypto bridge
 */
class MatrixCryptoBridge(
    userId: String,
    deviceId: String,
    pickleKey: String
) {
    private val crypto: MatrixCrypto = MatrixCrypto(
        userId = userId,
        deviceId = deviceId,
        pickleKey = pickleKey
    )

    /**
     * Get the device fingerprint
     */
    fun deviceFingerprint(): String = crypto.deviceFingerprint()

    /**
     * Get the user ID
     */
    fun userId(): String = crypto.userId()

    /**
     * Get the device ID
     */
    fun deviceId(): String = crypto.deviceId()

    /**
     * Get devices for a user
     */
    @Throws(CryptoException::class)
    fun getUserDevices(userId: String): List<DeviceInfo> =
        crypto.getUserDevices(userId)

    /**
     * Add a device to the store
     */
    @Throws(CryptoException::class)
    fun addDevice(device: DeviceInfo) {
        crypto.addDevice(device)
    }

    /**
     * Start device verification
     */
    @Throws(CryptoException::class)
    fun startVerification(
        otherUserId: String,
        otherDeviceId: String
    ): VerificationState = crypto.startVerification(
        otherUserId = otherUserId,
        otherDeviceId = otherDeviceId
    )

    /**
     * Get SAS emojis for verification
     */
    @Throws(CryptoException::class)
    fun getSASEmojis(verificationId: String): List<EmojiSASPair> =
        crypto.getSASEmojis(verificationId)

    /**
     * Confirm SAS verification
     */
    @Throws(CryptoException::class)
    fun confirmSAS(verificationId: String) {
        crypto.confirmSAS(verificationId)
    }

    /**
     * Complete device verification
     */
    @Throws(CryptoException::class)
    fun completeVerification(verificationId: String) {
        crypto.completeVerification(verificationId)
    }

    /**
     * Cancel device verification
     */
    @Throws(CryptoException::class)
    fun cancelVerification(verificationId: String) {
        crypto.cancelVerification(verificationId)
    }

    /**
     * Get verification state
     */
    @Throws(CryptoException::class)
    fun getVerificationState(verificationId: String): VerificationState =
        crypto.getVerificationState(verificationId)

    /**
     * Enable encryption for a room
     */
    @Throws(CryptoException::class)
    fun enableRoomEncryption(
        roomId: String,
        algorithm: String
    ) {
        crypto.enableRoomEncryption(roomId, algorithm)
    }

    /**
     * Get room encryption state
     */
    @Throws(CryptoException::class)
    fun getRoomEncryptionState(roomId: String): RoomEncryptionState =
        crypto.getRoomEncryptionState(roomId)

    /**
     * Encrypt event content
     */
    @Throws(CryptoException::class)
    fun encryptEvent(
        roomId: String,
        eventType: String,
        content: String
    ): String = crypto.encryptEvent(
        roomId = roomId,
        eventType = eventType,
        content = content
    )

    /**
     * Decrypt event content
     */
    @Throws(CryptoException::class)
    fun decryptEvent(
        roomId: String,
        encryptedContent: String
    ): String = crypto.decryptEvent(
        roomId = roomId,
        encryptedContent = encryptedContent
    )
}

/**
 * Exception thrown by crypto operations
 */
class CryptoException(message: String, cause: Throwable? = null) :
    Exception(message, cause)

/**
 * Information about a device
 */
data class DeviceInfo(
    val deviceId: String,
    val userId: String,
    val displayName: String? = null,
    val fingerprint: String,
    val isVerified: Boolean = false,
    val isBlocked: Boolean = false,
    val algorithm: String
)

/**
 * A pair of emoji for SAS verification
 */
data class EmojiSASPair(
    val emoji: String,
    val name: String
)

/**
 * The state of a device verification
 */
data class VerificationState(
    val verificationId: String,
    val state: String,
    val otherUserId: String,
    val otherDeviceId: String,
    val emojis: List<EmojiSASPair> = emptyList(),
    val decimals: List<Int> = emptyList()
)

/**
 * Room encryption state
 */
data class RoomEncryptionState(
    val roomId: String,
    val isEncrypted: Boolean,
    val algorithm: String? = null,
    val trustedDevices: List<String> = emptyList(),
    val untrustedDevices: List<String> = emptyList()
)

/**
 * Encryption algorithm information
 */
data class EncryptionAlgorithm(
    val algorithm: String,
    val rotationPeriodMs: Long,
    val rotationPeriodMsgs: Long
)

/**
 * User device list
 */
data class UserDevices(
    val userId: String,
    val devices: List<DeviceInfo>
)

File: matrix-crypto-ios/MatrixCryptoBridge.swift
import Foundation

/// Swift wrapper for the Matrix crypto bridge
public class MatrixCryptoBridge {
    private let crypto: MatrixCrypto
    
    /// Initialize the crypto bridge
    /// - Parameters:
    ///   - userId: The user ID (e.g., @user:example.com)
    ///   - deviceId: The device ID
    ///   - pickleKey: The encryption key for storage
    public init(userId: String, deviceId: String, pickleKey: String) throws {
        self.crypto = try MatrixCrypto(
            userId: userId,
            deviceId: deviceId,
            pickleKey: pickleKey
        )
    }
    
    /// Get the device fingerprint
    public func deviceFingerprint() -> String {
        return crypto.deviceFingerprint()
    }
    
    /// Get the user ID
    public func userId() -> String {
        return crypto.userId()
    }
    
    /// Get the device ID
    public func deviceId() -> String {
        return crypto.deviceId()
    }
    
    /// Get devices for a user
    public func getUserDevices(userId: String) throws -> [DeviceInfo] {
        return try crypto.getUserDevices(userId: userId)
    }
    
    /// Add a device to the store
    public func addDevice(_ device: DeviceInfo) throws {
        try crypto.addDevice(device)
    }
    
    /// Start device verification
    public func startVerification(
        otherUserId: String,
        otherDeviceId: String
    ) throws -> VerificationState {
        return try crypto.startVerification(
            otherUserId: otherUserId,
            otherDeviceId: otherDeviceId
        )
    }
    
    /// Get SAS emojis for verification
    public func getSASEmojis(verificationId: String) throws -> [EmojiSASPair] {
        return try crypto.getSASEmojis(verificationId: verificationId)
    }
    
    /// Confirm SAS verification
    public func confirmSAS(verificationId: String) throws {
        try crypto.confirmSAS(verificationId: verificationId)
    }
    
    /// Complete device verification
    public func completeVerification(verificationId: String) throws {
        try crypto.completeVerification(verificationId: verificationId)
    }
    
    /// Cancel device verification
    public func cancelVerification(verificationId: String) throws {
        try crypto.cancelVerification(verificationId: verificationId)
    }
    
    /// Get verification state
    public func getVerificationState(verificationId: String) throws -> VerificationState {
        return try crypto.getVerificationState(verificationId: verificationId)
    }
    
    /// Enable encryption for a room
    public func enableRoomEncryption(
        roomId: String,
        algorithm: String
    ) throws {
        try crypto.enableRoomEncryption(roomId: roomId, algorithm: algorithm)
    }
    
    /// Get room encryption state
    public func getRoomEncryptionState(roomId: String) throws -> RoomEncryptionState {
        return try crypto.getRoomEncryptionState(roomId: roomId)
    }
    
    /// Encrypt event content
    public func encryptEvent(
        roomId: String,
        eventType: String,
        content: String
    ) throws -> String {
        return try crypto.encryptEvent(
            roomId: roomId,
            eventType: eventType,
            content: content
        )
    }
    
    /// Decrypt event content
    public func decryptEvent(
        roomId: String,
        encryptedContent: String
    ) throws -> String {
        return try crypto.decryptEvent(
            roomId: roomId,
            encryptedContent: encryptedContent
        )
    }
}
